Docs for creating a virtual computer with LSfR (#339)
AlexsLemonade/OpenScPCA-analysis: docs/software-platforms/lsfr/creating-vcs.md

# Creating virtual computers with Lightsail for Research

While working on an [analysis module](../../contributing-to-analyses/analysis-modules/index.md), you can develop locally or on a virtual computer with Lightsail for Research (LSfR).

> Virtual computers provide access to a set amount of virtual CPUs, memory, and storage through a web browser rather than a physical machine.

Reviewer: ```suggestion
Virtual computers provide access to a set amount of virtual CPUs (vCPUs), memory, and storage through a web browser rather than a physical machine.
```

Using a virtual computer allows you to develop and run analyses that require more CPUs and memory than is available on your local machine.

_Note that you will need access to AWS to use LSfR._

> See [getting access to AWS](../../getting-started/accessing-resources/index.md#getting-access-to-aws) to learn how to get set up with AWS.

Reviewer: Make an actual note?

- The virtual computers provided through LSfR use an Ubuntu operating system.
- During setup, you will choose your desired configuration, including the amount of virtual CPUs, memory, and storage space.
    - You also have the option to add [additional storage to your virtual computer](STUB-LINK for volumes).
- All virtual computers will have a set of pre-installed applications needed for working with OpenScPCA, including:
    - Git and GitKraken
    - The AWS command line interface
    - R
    - Conda

> 

Reviewer: Note that they don't have to _configure_ AWS CLI. That will already be done!

> !!! note "More on virtual computers"

Reviewer: Tip?

    Read more about [virtual computers with Lightsail for Research](https://docs.aws.amazon.com/lightsail-for-research/latest/ug/computers.html).

## How to create a virtual computer

Follow the below steps to create a virtual computer to use with LSfR.

1. Navigate to the [access portal URL from when you set up your user in IAM Identity center](../aws/index.md#joining-iam-identity-center).
This link should bring you to the AWS Console and home page.

    Once here, be sure that you are in the `us-east-2` region, by selecting the drop-down menu in the tool bar.

    <figure markdown="span">
        ![Select region](../../img/creating-vcs-1.png){width="400"}
    </figure>

> 1. You will need to use the AWS Service Catalog to create virtual computers.

Reviewer: We should note that directly creating an instance via Lightsail for Research is not supported. Mostly raising this because we're linking to LfR docs.

Open the AWS Service Catalog by using the search bar and typing, `service catalog`.

    <figure markdown="span">
        ![Select service catalog](../../img/creating-vcs-2.png){width="600"}
    </figure>

1. Select `LightsailInstance` in the product list, and click `Launch product`.

    <figure markdown="span">
        ![Launch product](../../img/creating-vcs-3.png){width="600"}
    </figure>

1. You will then choose the name and configurations for your virtual computer.

    - Start by providing a `Provisioned product name`.

>     <!--TODO Do we want to provide guidance on names?-->

Reviewer: I tend to name it the same thing as the instance name to make it easier to find which provisioned product corresponds to which virtual computer. (Though I don't expect researchers to do much of that necessarily).

    - Pick an `Application` to use from the drop-down menu.
    This installs any additional applications along with the pre-installed applications.
        - If you are planning to develop in R, we recommend choosing `Rstudio`, which includes RStudio.
        - If you are planning to develop in Python, we recommend choosing `VSCodium`, which includes VSCodium, a distribution of Microsoft's editor VS Code.
        - For information on all application options, see [the LSfR documentation on Applications](https://docs.aws.amazon.com/lightsail-for-research/latest/ug/blueprints-plans.html).

>     - Choose any of the availability zones.

Reviewer: I think we might need to recommend `us-east-2a` because if your volume is not in the same AZ (which is not at all transparent to you as a user!), you can't attach it, which does make sense to me.

We should probably remove the other options eventually.

>     - Name your instance, be sure to make sure this is different than the name chose for your product.

Reviewer: Did you receive an error that made you make this recommendation or is it because the default name is not descriptive? (I do see your TODO.)

    - Pick the size for your instance.

>         - Use this [table describing the total vCPUs, memory, and storage space](https://docs.aws.amazon.com/lightsail-for-research/latest/ug/blueprints-plans.html#plans) to choose the most appropriate instance plan.

Reviewer: We need to recreate this table here to report pricing in `us-east-2`. See https://data-lab-knowledge.slab.com/posts/2024-03-23-jnt-notes-from-testing-researcher-aws-access-g0sdd98p#hnkio-information-only-section. I grabbed these from the LfR console. When you create via the LfR console, they will tell you how much things cost. I would also reiterate that they have a monthly budget and link to https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/getting-started/accessing-resources/getting-access-to-compute.md#monthly-budget.

        - Before you request a plan with GPUs, you will need to put in a [request with the Data Lab team](../../getting-started/accessing-resources/getting-access-to-compute.md#gpu-instance-access).
    - Set the `ShutdownIdlePercent` to 5 and the `ShutdownTimePeriod` to 10.
    This means if the instance is using less than 5% of the total available CPUs or inactive for more than 10 minutes, the instance will temporarily shut down.

>     This _does not_ terminate your instance, it temporarily stops any idle instances until you are ready to resume your work.

Reviewer: This seems less jargony to me.
```suggestion
    This _does not_ delete your instance; it temporarily stops any idle instances until you are ready to resume your work.
```

    <figure markdown="span">
        ![Configure instance](../../img/creating-vcs-4.png){width="600"}
    </figure>

1. Once you have configured your instance, click `Launch product`.
You have now created a virtual computer!

## How to access a virtual computer

Once you have created your virtual computer, you will be able to access and launch the computer on a web browser.

> Note that it can take 5-10 minutes to create the instance.

Reviewer: In my experience, you can access an instance before GitKraken and R installation is completed. I think you might have encountered that in testing based on your notes.
```suggestion
We recommend waiting ~10 minutes before accessing the instance.
It takes time create the instance and for all software to be installed.
```

1. To access a virtual computer, you will need to navigate to Lightsail for Research.
Search for Lightsail for Research using the search bar and then select the product.

    <figure markdown="span">
        ![Search for Lightsail](../../img/creating-vcs-5.png){width="600"}
    </figure>

1. This will take you directly to a page that lists your virtual computers.
You should see the virtual computer you created with the instance name that you provided during set up.

    <figure markdown="span">
        ![List of computers](../../img/creating-vcs-6.png){width="600"}
    </figure>

1. To launch the computer, click `Start computer`.
When the computer is ready, use the drop-down menu in the lower right-hand side to select `Access operating system`.

    <figure markdown="span">
        ![Start computer](../../img/creating-vcs-7.png){width="600"}
    </figure>

>     You can also use `Launch Rstudio` or `Launch VSCodium`, but if you do that you will have access to _only_ that application instead of the Ubuntu desktop and all installed applications.

Reviewer: Something I experienced when opening RStudio is that I had to hit the `x` to close it before it pointed to 4.3.3. Might want to note that here? Or maybe we create a troubleshooting section for LfR this week?

Author: Hmm I was able to open Rstudio both from clicking `Launch Rstudio` and from opening the application on the desktop and R 4.3.3 was being used?

Reviewer: I think it boils down to timing.

# Creating virtual computers with Lightsail for Research

While working on an [analysis module](../../contributing-to-analyses/analysis-modules/index.md), you can develop locally or on a virtual computer with Lightsail for Research (LSfR).

Virtual computers provide access to a set amount of virtual CPUs (vCPUs), memory, and storage through a web browser rather than a physical machine.
Using a virtual computer allows you to develop and run analyses that require more CPUs and memory than is available on your local machine.

!!! note

    You will need access to AWS to use LSfR.
    See [getting access to AWS](../../getting-started/accessing-resources/index.md#getting-access-to-aws) to learn how to get set up with AWS.

> - The virtual computers provided through LSfR use an Ubuntu operating system.

Reviewer: ```suggestion
- The virtual computers provided through LSfR use the Ubuntu operating system.
```

- During setup, you will choose your desired configuration, including the amount of virtual CPUs, memory, and storage space.

>     - You also have the option to add [additional storage to your virtual computer](STUB-LINK for volumes).

Reviewer: ```suggestion
    - You also have the option to add [additional storage to your virtual computer](STUB_LINK for volumes).
```

Author: I swear I did this but didn't commit it. Thank you.

- All virtual computers will have a set of pre-installed applications needed for working with OpenScPCA, including:
    - Git and GitKraken
    - The AWS command line interface, which has already been configured for you
    - R
    - Conda

!!! tip "More on virtual computers"

    Read more about [virtual computers with Lightsail for Research](https://docs.aws.amazon.com/lightsail-for-research/latest/ug/computers.html).

## How to create a virtual computer

Follow the below steps to create a virtual computer to use with LSfR.

1. Navigate to the [access portal URL from when you set up your user in IAM Identity center](../aws/index.md#joining-iam-identity-center).

> This link should bring you to the AWS Console and home page.

Reviewer: This is missing a step, I think – don't you need to select the account and then click on `ResearcherRestriction` to get into the console?

1. To access a virtual computer, you will need to navigate to Lightsail for Research.
Search for Lightsail for Research using the search bar and then select the product.

    <figure markdown="span">
        ![Search for Lightsail](../../img/creating-vcs-5.png){width="600"}
    </figure>

1. This will take you directly to a page that lists your virtual computers.
You should see the virtual computer you created with the instance name that you provided during set up.

    <figure markdown="span">
        ![List of computers](../../img/creating-vcs-6.png){width="600"}
    </figure>

1. To launch the computer, click `Start computer`.
When the computer is ready, use the drop-down menu in the lower right-hand side to select `Access operating system`.

    <figure markdown="span">
        ![Start computer](../../img/creating-vcs-7.png){width="600"}
    </figure>

    You can also use `Launch Rstudio` or `Launch VSCodium`, but if you do that you will have access to _only_ that application instead of the Ubuntu desktop and all installed applications.

1. A new window should open in your browser with the Ubuntu desktop view.

> You can now continue to [develop analysis using your virtual computer](STUB_LINK to developing with LSfR).

Reviewer: ```suggestion
You can now continue to [develop analyses using your virtual computer](STUB_LINK to developing with LSfR).
```
